Introduce FormField\ScopeBuilder

georgehristov opened the pull request:
Introduces GUI for setting up Model scopes using [VueQueryBuilder](https://dabernathy89.github.io/vue-query-builder/)

Below is based on demos `Stat` model

![scope](https://user-images.githubusercontent.com/8128250/80114118-75b18680-8583-11ea-81c9-85222ce5464e.png)

georgehristov committed:
introduce FormField\ScopeBuilder

romaninsh committed:
Apply fixes from StyleCI

georgehristov committed:
enable list delimiter auto-detection

romaninsh committed:
Apply fixes from StyleCI

romaninsh commented:
I'd like to hear what others think - should this be in this repo or as add-on?

georgehristov committed:
use scope builder as atkComponent

mvorisek commented:
> 
> 
> I'd like to hear what others think - should this be in this repo or as add-on?

It is quite specifc, I think this repo should provide only basic and well maintained UI components and complex components only if they are used a lot and enought flexible. So I am for an extra repo.

If this component will be used a lot, it can be always moved here.

What about a 2nd repo like `ui-extra` as a central place for early / unmature components?

georgehristov commented:
In fact I would also agree this is not part of the base UI components and it would be better off in a separate repo.
Then we just need to collect opinions if stand-alone or as a bundle (`ui-extra` or `ui-extended` or ...)

mvorisek commented:
I prefer bundle/addon but with demos like in this repo. CI scripts can be copied 1:1 (incl. phpunit and Behat testing). But it should be extra, not extended. (but feel free still to propose a better name)

romaninsh commented:
I suppose the difference is - if this component in here, then any PRs into core repositories (atk data for example), will need to make condition builder is updated. If it's in a separate repository then @georgehristov would have to maintain it himself. 

I can see where scope builder could be used inside column filters, so that's an argument to include builder here. This would also allow other add-ons to rely on it easily.

mvorisek commented:
> I can see where scope builder could be used inside column filters, so that's an argument to include builder here.

makes sense to include here for another argument - it is only one file

ibelar commented:
If we include it in UI than I would propose adding all js related files into atkjs bundle. This way, we would not have to worry about including them into pages and it should work in Modal and VirtualPage out of the box. 

@georgehristov - I am working into creating a Fomantic-UI look and feel for this component and already have extended it using the Fomantic CSS template. I can take care of that part also if you want. Plus we need a calendar component for date field in it, so it could be included here as well. The calendar component could be used in MultiLine too.

georgehristov commented:
@ibelar Integrating this with FUI sounds great and really appreciated!

georgehristov commented:
@ibelar I have included it in atkjs in `atkComponents` as `atk-query-builder` but have not pushed this commit.
Shall I push it for reference or you are further down the road than this already?

georgehristov committed:
revert inclusion of css

ibelar committed:
Merge branch 'develop' into feature/formfield-scope-bulder

mvorisek commented on js/src/services/vue.service.js:
```diff
@@ -15,6 +16,7 @@ let atkComponents = {
   'atk-item-search' : itemSearch,
   'atk-multiline'   : multiLine,
   'atk-tree-item-selector' : treeItemSelector,
+  'atk-query-builder' : VueQueryBuilder,
```
@ibelar why is this needed? Do we need to change js for every Vue based component added?

mvorisek commented on src/FormField/ScopeBuilder.php:
```diff
@@ -0,0 +1,576 @@
+     * @var array
+     */
+    protected static $operators = [
+        'equals' => '=',
```
@georgehristov you use these natural strings on multiple locations, can you declare them in constants once and then use them?

Example:
```
public const OPERATOR_DOES_NOT_EQUAL = 'does not equal';

something(self::OPERATOR_DOES_NOT_EQUAL);
```

georgehristov replied:
@mvorisek I also like using constants for definition but the challenge here is that we have operator label and operator value. Will think about how to implement it.

mvorisek replied:
![image](https://user-images.githubusercontent.com/2228672/84356754-f0327600-abc4-11ea-8548-c8021feaa8e5.png)

something that is unique... it needs to be probably the name/label :)

georgehristov committed:
use class constants for operators instead of text

romaninsh committed:
Apply fixes from StyleCI

mvorisek commented on src/FormField/ScopeBuilder.php:
```diff
@@ -524,16 +543,16 @@ public static function conditionToQuery(Condition $condition): array
-                    'contains'
+                    self::OPERATOR_EQUALS,
+                    self::OPERATOR_BEGINS,
+                    self::OPERATOR_ENDS,
```
`ENDS_WITH` sound much more natural

georgehristov replied:
Done now, I am a proponent of good semantics in code

mvorisek replied:
Great!

mvorisek commented on src/FormField/ScopeBuilder.php:
```diff
@@ -524,16 +543,16 @@ public static function conditionToQuery(Condition $condition): array
+                    self::OPERATOR_DOESNOT_EQUAL,
+                    self::OPERATOR_DOESNOT_BEGIN,
+                    self::OPERATOR_DOESNOT_END,
+                    self::OPERATOR_DOESNOT_CONTAIN,
```
`DOESNOT` sounds not ideal - is `NOT_CONTAINS` better?

georgehristov replied:
I tried it but I did not like it

mvorisek replied:
You marked it `protected` that I think it is ok... I do not like boths but have no better now :D

mvorisek commented on src/FormField/ScopeBuilder.php:
```diff
@@ -549,7 +568,7 @@ public static function conditionToQuery(Condition $condition): array
                 $value = implode(',', $value);
                 $operator = $map[$operator] ?? 'IN';
```
missed this? 3 maybe there are even more placed...

georgehristov replied:
Not really as this is the `atk4\data\Scope` version of operators.
Constants in the class represent the VueQueryBuilder operators

georgehristov committed:
improve semantics

mvorisek commented on src/FormField/ScopeBuilder.php:
```diff
@@ -72,30 +72,49 @@ class ScopeBuilder extends Generic
-        'is empty' => '=',
-        'is not empty' => '!=',
+        self::OPERATOR_EQUALS => '=',
+        self::OPERATOR_DOESNOT_EQUAL => '!=',
```
Does `negative` operators matches nulls or is there any option to set it like so - we should have real inverse operator.

georgehristov replied:
Can you give an example as I do not fully understand what you mean?

mvorisek replied:
table cell with value `null` is not matched with `= 1` but also NOT matched with `!= 1` (at least in MySQL/Maria)

georgehristov replied:
Now operators `empty`, `not empty` result in checking against `NULL` value

mvorisek replied:
![image](https://user-images.githubusercontent.com/2228672/88925942-8dc92e00-d275-11ea-819a-383012d93f73.png)

@georgehristov please file whole ui

then what feedback is left for this PR?

georgehristov replied:
Fixed, I believe it is RTM

I will use in in atk4/login and we'll see if any polishing necessary

mvorisek commented on src/FormField/ScopeBuilder.php:
```diff
@@ -0,0 +1,595 @@
+     *
+     * @var integer
+     */
+    public $maxDepth = 10;
```
Have this any performance side effect on ui side if large? (test with 10000)

If not, let's set at least to 25.

georgehristov replied:
No performance issues but this will be a hell of a Scope 😄 
I have never used more than 3 - 4 sub-levels.
Not sure if you got what this setting means.
You can always set a value when creating the ScopeBuilder object in your code

mvorisek replied:
But in rare cases you may need 11, not that far from 3-4 :D

mvorisek commented on src/FormField/ScopeBuilder.php:
```diff
@@ -0,0 +1,595 @@
+     *
+     * @var array
+     */
+    public static $listDelimiters = [';', ',', '|'];
```
When `|` makes sense? I think we do not need it at all.

georgehristov replied:
Aiming at more versatility. Some users are used to using | for a list delimiter (from DynamicsNAV as one example)

mvorisek replied:
I think better to not make default because it is confusing with `||`.

mvorisek replied:
@georgehristov can we remove `|` from defaults?

georgehristov replied:
Done

mvorisek committed:
Fix refactored hook name

mvorisek committed:
fix CS

mvorisek commented on src/FormField/ScopeBuilder.php:
```diff
@@ -0,0 +1,565 @@
+        switch ($type) {
+            case 'query-builder-group':
+                $components = array_map([static::class, 'queryToScope'], $query['children']);
+                $junction = $query['logicalOperator'] === 'all' ? Scope::AND : Scope::OR;
```
I do not like `all` name

mvorisek replied:
![image](https://user-images.githubusercontent.com/2228672/84361748-65ee1000-abcc-11ea-9f88-0f036e4c6152.png)

In GUI we should have definitely `OR` and/or `AND` - people are used to use these even from non-SQL world

mvorisek replied:
should be also updated in data/scope: `scope->all` is simply not ok

georgehristov replied:
Changed

georgehristov committed:
remove throws from comments

ibelar committed:
Merge branch 'develop' into feature/formfield-scope-bulder

ibelar committed:
webpack config and bundle update

ibelar committed:
test segment template

mvorisek committed:
Merge branch 'develop' into feature/formfield-scope-bulder

ibelar committed:
using card

ibelar committed:
update styling

ibelar committed:
Merge branch 'feature/formfield-scope-bulder' of https://github.com/atk4/ui into feature/formfield-scope-bulder

ibelar committed:
Merge branch 'feature/formfield-scope-builder-with-segment' into feature/formfield-scope-bulder

ibelar committed:
remove hard copy data + minor fix

ibelar committed:
compnent testing

ibelar committed:
update package info

ibelar committed:
cs fixer

mvorisek commented on demos/input/query-builder.php:
```diff
@@ -0,0 +1,24 @@
+
+$f = \atk4\ui\Form::addTo($app);
+$qb = $f->addField('qb', [\atk4\ui\FormField\ScopeBuilder::class]);
+$qb->rules = $rules;
```
@ibelar can you setup your IDE to always add end EOL for text files on save? Seen this many times and for pug/html it is currently not checked by our CI tests.

ibelar committed:
map input value in component

mvorisek commented on demos/input/query-builder.php:
```diff
@@ -22,3 +22,7 @@
 $qb->rules = $rules;
+
+$f->onSubmit(function($f) {
+   echo  '<pre>' . json_encode($f->model->get('qb'), JSON_PRETTY_PRINT) . '</pre>';
```
Use `App->jsonEncode` (if not only for some test...)

ibelar committed:
add more html input type

ibelar committed:
set renderLabel false by default

ibelar committed:
Merge branch 'develop' into feature/formfield-scope-bulder

ibelar commented:
@georgehristov - Does scope in data support operator 'BETWEEN'? If so, I can add an extra input to the query-builder when the user selects 'BETWEEN' as the operator. As of now, the name of the parameter for the first input is 'value'. What should be the extra one? (value_2, extra)

georgehristov commented:
@ibelar BETWEEN is not supported in Model\Scope.
Implementing it is too complicated and functionality is there by using 2 simple conditions joined by AND

ibelar commented:
@georgehristov - Ok. I will probably still add it to UI then and generate proper AND query when user select ‘BETWEEN’. It make sense on a user stand point to use it.

georgehristov commented:
@ibelar that is what I was going to say. If we want to introduce to UI then we should map it to 2 conditions.
Challenge here is to map it 2 way - UI to scope and scope to UI

ibelar committed:
Merge branch 'develop' into feature/formfield-scope-bulder

georgehristov committed:
[update] switch to new method name

ibelar committed:
add date rule

- use v-date-picker for date

ibelar committed:
Merge branch 'develop' into feature/formfield-scope-bulder

ibelar committed:
Merge branch 'develop' into feature/formfield-scope-bulder

ibelar committed:
imorvement

- better ui
- improve minimizing js using terser plugin
- set default at vue-query-builder component level.

ibelar committed:
move button to left

georgehristov commented on src/FormField/ScopeBuilder.php:
```diff
@@ -28,10 +29,11 @@ class ScopeBuilder extends Generic
     /**
      * Max depth of nested conditions allowed.
      * Corresponds to VueQueryBulder maxDepth.
+     * Maximum support by js component is 5.
```
@ibelar This limit would be too low.
Any reason to set such hardcoded limit?

ibelar replied:
The marketing department suggests starting with a lower limit. We can promote it as a new feature in the next release. :D 
Seriously no specific reason. What would be a good limit?

georgehristov replied:
:D

I have used 5-6 levels so that's why I set it to 10.
@mvorisek proposed 25 but this would be a hell of a scope

ibelar replied:
Let's start with 10. We can always readjust if we need more.

ibelar committed:
use basic button

ibelar committed:
Merge branch 'develop' into feature/formfield-scope-bulder

ibelar committed:
update radio

ibelar committed:
update dockerfile for latest node-npm

mvorisek commented on demos/Dockerfile:
```diff
@@ -13,10 +13,16 @@ RUN apt-get update && apt-get install -y \
 RUN curl -sS https://getcomposer.org/installer | php -- --install-dir=/usr/local/bin --filename=composer
 
+# update npm - apt-get is not installing the latest version.
+RUN npm install -g npm
```
should be put probably after install nodejs

ibelar replied:
As long as it is before doing any installation. Besides, I have tested it and it was ok.

ibelar committed:
update npm

ibelar committed:
test

mvorisek commented on .github/workflows/unit-tests.yml:
```diff
@@ -58,7 +58,7 @@ jobs:
         if: matrix.type == 'Behat'
         run: |
-          apk add npm
+          apk add --update npm
```
simple, nice :)

ibelar replied:
tks to Google! :)

ibelar committed:
update travis

ibelar committed:
more travis test

ibelar committed:
bump node js

ibelar committed:
test

ibelar committed:
test

ibelar committed:
install curl

ibelar committed:
switch to bionic

ibelar committed:
remove js

ibelar committed:
add checkbox

ibelar committed:
add comments

ibelar commented:
@georgehristov 

This is the final look. It now supports Date, Radio, Checkbox plus regular input of course. You can see how to rule is set up for them in this temp file: /demos/input/query-builder.php

P.S. The date picker relies on v-date-picker so we can use some of its options there.

Are you ok to hook it up to data/scope? I did not add between yet. I will wait to see how it looks like once set via a model.

<img width="750" alt="Screen Shot 2020-06-17 at 1 59 17 PM" src="https://user-images.githubusercontent.com/2204478/84933009-3135f900-b0a3-11ea-80dc-85f242149c15.png">

mvorisek commented on js/src/components/query-builder/query-builder.component.vue:
```diff
@@ -33,8 +30,9 @@
         rules: this.data.rules ? this.data.rules : [],
         name: this.data.name ? this.data.name : '',
-        maxDepth: this.data.maxDepth ? ((this.data.maxDepth <= 5) ? this.data.maxDepth : 5 ): 1,
+        maxDepth: this.data.maxDepth ? ((this.data.maxDepth <= 10) ? this.data.maxDepth : 10 ): 1,
```
why limit on two places?

ibelar replied:
Just a safety on the component side.

mvorisek replied:
So with 100 the frontend will catch on fire? :D

mvorisek replied:
I think this safety check should be removed (it checks if data from php are ok?) and fallback to 10 if not specified. Simple.

ibelar replied:
because the query can come from anywhere. If stored in DB for example, we may not have control over it. I prefer to limit it here and 10 levels should be quite enough.

georgehristov commented:
@ibelar Wonderful! I will take it up for hooking to data/scope.

mvorisek commented:
![image](https://user-images.githubusercontent.com/2228672/84943704-0d93a280-b0e5-11ea-81ec-02b2b082a50d.png)

@ibelar look very good

What about adding a little (1/3 - 1/2) space between the rules?

Also, the "X"es are very dominant and distracting. Can we display them only when a mouse is over that rule or show it with light grey color only when out of focus? (both can be can easily in CSS)

georgehristov committed:
[update] link with model scope

ibelar commented:
> What about adding a little (1/3 - 1/2) space between the rules?

I can propose one version with some space in it. Default spacing was definitely too large, but may be some space will be better.

> Also, the "X"es are very dominant and distracting. Can we display them only when a mouse is over that rule or show it with light grey color only when out of focus? (both can be can easily in CSS)

You are right, I will make it smaller. I do no like hover because it is not that intuitive. Plus, hover on a phone or tablet is not that great.

georgehristov committed:
[update] add to demo menu

georgehristov commented:
@ibelar Now I have "linked" it with Model scope and updated the demo.

You need to require `"atk4/data" : "dev-feature/model-scope"` in composer.json

Some issues I noticed:
- multiple `Duplicate keys` error in console
- "Add Rule" dropdown can get quite long - maybe make scrollable?
- radio selects for "Is Commercial" field select both "Yes" and "No" on click
- I know it is default functionality in VueQueryBuilder but it is annoying that when user selects "is empty" option the input for text is still shown. Can you do something about it?
- "Currency" field which is enum type does not show a dropdown select

georgehristov commented:
> What about adding a little (1/3 - 1/2) space between the rules?

I find no spaces really good as it visually separates sub-scopes. In any case space should be very small for that purpose

ibelar committed:
Merge branch 'feature/formfield-scope-bulder' of https://github.com/atk4/ui into feature/formfield-scope-bulder

ibelar committed:
vue component changed

- add option for spacing between rule and group
- add option for changing remove icon for rule and group
- reduce size of remove icon
- set lighter color on remove icon

mvorisek commented on js/src/components/query-builder/query-builder.component.vue:
```diff
@@ -41,20 +41,27 @@
+          addGroup: "Add Group",
+          removeGroup: "small icon times",
+          textInputPlaceholder: "value",
+          spaceRule: "fitted", // can be compact or padded
```
so can it be fitted?

ibelar replied:
yes - good point!

ibelar commented:
Component is update 

I have reduced the size of the remove icon and change it's color so it is a lot less disturbing. I also add the possibility to add space in between rule via the component labels options. The rule is called spaceRule and you have a choice of 'fitted', 'compact', or 'padded'. Default is fitted, i.e. no space and 'padded' is the maximum space.

Here is a look with 'compact' option using new remove icon.

<img width="894" alt="Screen Shot 2020-06-17 at 5 50 36 PM" src="https://user-images.githubusercontent.com/2204478/84954682-d9f45080-b0c3-11ea-87da-b538d5df2329.png">

mvorisek commented:
@ibelar ❤️ it, X should be "same light gray" as button text "Add Rule"

one more question - is that really that huge in reality? Based on your screen, there only like 8 rules fits 1200p screen/non-fullscreen browser!

it should no larger/higher than this: (click for real size)
![image](https://user-images.githubusercontent.com/2228672/84955340-6e2bdb00-b0f7-11ea-8b0e-8abd56b21ce8.png)

ibelar commented:
> one more question - is that really that huge in reality? Based on your screen, there only like 8 rules fits 1200p screen/non-fullscreen browser!

It uses the Fomantic card component with fluid class. It will be as large as the container you are placing it in too. You can easily change that via `scopeBuilderTemplate` property.

ibelar commented:
> Some issues I noticed:
> 
> * multiple `Duplicate keys` error in console
> * "Add Rule" dropdown can get quite long - maybe make scrollable?
> * radio selects for "Is Commercial" field select both "Yes" and "No" on click
> * I know it is default functionality in VueQueryBuilder but it is annoying that when user selects "is empty" option the input for text is still shown. Can you do something about it?
> * "Currency" field which is enum type does not show a dropdown select

Yes - I will look into it.

mvorisek commented:
@ibelar I mean the current style is unusable - it is brutally big, like only 8-10 rules fit to the whole screen. We need like 40 (or in another words 1/3 of the height for header/footer/page title, 1/3 for some more content like manual, data, etc., and then within 1/4 - 1/3 of screen height the typical scope rules must fit.

**In other words, we have about 250px of height and 10 rules must fit in this space at least.**

Width is ok.

ibelar committed:
fix dropdown

ibelar commented:
> @ibelar I mean the current style is unusable - it is brutally big, like only 8-10 rules fit to the whole screen. We need like 40 (or in another words 1/3 of the height for header/footer/page title, 1/3 for some more content like manual, data, etc., and then within 1/4 - 1/3 of screen height the typical scope rules must fit.

Height is ok as it is. Reducing the height of each rule would only make it harder to read. That is why I remove space in-between rules so we can fit more of them. Can you show me a screen sample of what you are saying?

mvorisek commented:
@ibelar click your and my screenshots above - my is 1:1, your probably too and compare. Your is simply too big. This is simply not acceptable for any serious/nonsimple/backoffice rules.

ibelar commented:
> @ibelar click your and my screenshots above - my is 1:1, your probably too and compare. Your is simply too big. This is simply not acceptable for any serious/nonsimple/backoffice rules.

@mvorisek - My mistake, I meant a screen shot of what you would really want. Did you see something similar you would like to share. I can see if I can adjust.

mvorisek commented:
@ibelar see my comments - if you posted 1:1 (ie. with 100% zoom), then the height of every element needs to be reduces to fix about 30 - 40 rules to standard 1080 px height monitor.

ibelar commented:
@ mvorisek - The actual look is very acceptable and no more change will be done on it. Only debugging ui glitch will be done at this stage.

mvorisek commented:
@ibelar the example from you is 825 px height with only 5 rules (4 rules + 1 nesting). @DarkSide666 @georgehristov do you really think this is usable?

Let's take an alternative fact. Excel. The world 1st data editor. When you open it with default settings/zoom, 35 rows fits within the same height.

So your score on this is only 14.3% (5/35) 😢

georgehristov commented:
@mvorisek Have you tested it?

It is quite fine from my point of view. Sizes appear very UI friendly - not too big, not too small.

Below is a screenshot on normal screen Agile UI demo page

![scope-builder](https://user-images.githubusercontent.com/8128250/85002753-9518fa00-b155-11ea-8c8a-90b2cce371d5.png)

mvorisek commented:
No. **I kindly asked if provided screen is 1:1**. Now/in your screen there are 17 rows. I still think the grey box paddings can be reduced, but it seems much better. (I propose 60 px, ie. not higher than the save button below)

And the value input field should be probably streched (ie. occupy all free width).

mvorisek commented:
Also, I do think display "==" is much better than displaying "equals" - much better UX? You do not trust me? Check some leading softwares? :D :D

So I propose to change the operator text to "==" (and >=, !=, ...) or "== (equals)"-

And match selectbox from "And" to "And (all rules must be satisfied)" (and "Or (At least one rule much be satisfied)") (this must be tested, I am not exactly sure if it is really that better)

georgehristov committed:
[update] demo model caption

ibelar committed:
fix duplicate id warning

ibelar committed:
Merge branch 'feature/formfield-scope-bulder' of https://github.com/atk4/ui into feature/formfield-scope-bulder

ibelar committed:
hide input on depending on operator

ibelar committed:
fix radio type

- value should be string

georgehristov committed:
[update] introduce inputType specific operators

georgehristov commented:
@ibelar I introduced some updates in PHP side
- debug is now an option that can be used any time (switched on for demo)
- inputType specific operators (`>`,`<`,`=`, etc for numeric, `before`, `after`, `on`, etc for dates)

I hope it does not interfere on your side

So @mvorisek we are now in line with "some leading softwares", breathing down their necks 😄

ibelar committed:
add select type

ibelar committed:
Merge branch 'feature/formfield-scope-bulder' of https://github.com/atk4/ui into feature/formfield-scope-bulder

mvorisek commented:
> inputType specific operators (>,<,=, etc for numeric, before, after, on, etc for dates)

why, how is before different to "<"?

> So @mvorisek we are now in line with "some leading softwares" 😄

@georgehristov your statement or measured result? If result, please provide data! I did :)

georgehristov committed:
[fix] queryToCondition

ibelar committed:
fix css glitch

ibelar commented:
@georgehristov - All should be working for the js component. Let me know if you see anything else.

ibelar committed:
minor fix

- update package dependency
- fix operator duplicate key
- fix date setting

mvorisek committed:
Cache file templates in memory (#1301)

mvorisek committed:
Add phpdoc to anonymous classes in demos (#1303)

ibelar committed:
add profile option

ibelar committed:
Merge branch 'develop' into feature/formfield-scope-bulder

georgehristov commented:
@ibelar Works great!

I just cannot figure out why tests failing (after I added query builder menu link for the demo)

Tests are running unbelievably slow on Windows

mvorisek committed:
Simplify Node update/install for Docker

mvorisek commented on demos/Dockerfile:
```diff
@@ -13,6 +13,10 @@ RUN apt-get update && apt-get install -y \
 
+RUN curl -sL https://deb.nodesource.com/setup_12.x  | bash - \
+    && apt-get update && apt-get -y install nodejs \
+    && npm install -g npm
```
@ibelar `npm` is installed above (line `libicu-dev git jq unzip npm`) - so move `npm` here after the `apt` sources was changed and remove `npm install -g npm`?

ibelar replied:
this is not for installing it but for updating it. It needs to be installed first. But not sure why it installed the wrong version. Probably because the image contains node v8

mvorisek replied:
The dist of the used image is conservative... After the sources are updated, I would expect we can install latest straight as I have advised above. I do not see any advantage why to install not-wanted version first.

georgehristov committed:
[fix] run CS Fixer

georgehristov commented:
Test fails as https://github.com/atk4/data/pull/551 is not merged yet.

@mvorisek Tests run much faster now!

mvorisek commented:
> @mvorisek Tests run much faster now!

yes, and they now also discover all demos and test them!

ibelar committed:
fix - date

ibelar committed:
adding bundle profile tools

ibelar committed:
update setModel for easier class extension

georgehristov committed:
Merge branch 'develop' into feature/formfield-scope-bulder

ibelar committed:
Merge branch 'develop' into feature/formfield-scope-bulder

ibelar committed:
update to latest dev

ibelar committed:
Merge branch 'develop' into feature/formfield-scope-bulder

ibelar committed:
update to latest develop

georgehristov committed:
Merge branch 'develop' into feature/formfield-scope-bulder

# Conflicts:
#	.travis.yml
#	demos/form-control/query-builder.php
#	js/package.json

georgehristov committed:
Merge branch 'feature/formfield-scope-bulder' of https://github.com/atk4/ui into feature/formfield-scope-bulder

# Conflicts:
#	demos/form-control/query-builder.php
#	js/package.json

georgehristov committed:
[update] use latest scope functionality

georgehristov commented:
@ibelar I have now updated this to reflect latest in atk4/data as scopes are merged now. The scope builder form control in demo however does not seem to submit data to backend. Can you pls check this?

georgehristov committed:
[update] include explicit empty operator to text, numeric, date fields and handle it

georgehristov commented:
This [Vue Query Builder](https://www.syncfusion.com/vue-ui-components/vue-query-builder) I find better but unfortunately it is paid 😄 
UI is closer to the originally used [jQuery QueryBuilder](https://querybuilder.js.org/)

mvorisek commented on src/Form/Control/ScopeBuilder.php:
```diff
@@ -538,7 +544,7 @@ public static function conditionToQuery(Scope\Condition $condition, $inputsMap =
         $value = $condition->value;
 
-        if (stripos($operator, 'like') !== false) {
+        if (str_contains($operator, 'like')) {
```
no, we should test for both `like`

mvorisek commented on src/Form/Control/ScopeBuilder.php:
```diff
@@ -0,0 +1,611 @@
+     *
+     * @var int
+     */
+    public $maxDepth = 3;
```
at least 5 please and then see if more is normally needed - user should never be limited!

ibelar committed:
fix

- set default query for component
- set form attribute name to input.

ibelar committed:
Merge branch 'feature/formfield-scope-bulder' of https://github.com/atk4/ui into feature/formfield-scope-bulder

ibelar committed:
merge

ibelar committed:
fix package version

ibelar committed:
use formElement name

georgehristov committed:
[update] use Condition operator constants

georgehristov committed:
[update] reduce default list delimiters

georgehristov commented:
@ibelar shall we merge this?

ibelar commented:
> shall we merge this?

@georgehristov - test is now failing:

1) atk4\ui\tests\DemosTest::testDemosStatusAndHtmlResponse with data set #40 ('form-control/query-builder.php')
31
Method Reference::getModel is deprecated. Use Model::getTheirModel instead

any idea?

georgehristov committed:
[fix] use new Reference::getTheirModel method

georgehristov commented:
@ibelar Fixed - reflecting latest updates to atk4/data

ibelar reviewed (approved):
Super!
